modin/experimental/core/execution/native/implementations/hdk_on_native/io/io.py

```diff
@@ -128,22 +129,54 @@ def read_csv(cls, **kwargs):  # noqa: PR01
                 raise ArrowEngineException(error_message)
 
-            dtype = kwargs["dtype"]
-            if isinstance(dtype, dict):
-                column_types = {c: cls._dtype_to_arrow(t) for c, t in dtype.items()}
-            else:
-                column_types = cls._dtype_to_arrow(dtype)
-
-            parse_dates = kwargs["parse_dates"]
-            if (type(parse_dates) is list) and type(column_types) is dict:
-                for c in parse_dates:
-                    column_types[c] = pa.timestamp("s")
-
-            names = kwargs["names"]
+            if (names := kwargs["names"]) is lib.no_default:
+                names = None
             skiprows = kwargs["skiprows"]
-            if names not in (lib.no_default, None) and kwargs["header"] == 0:
+            if names and kwargs["header"] == 0:
                 skiprows = skiprows + 1 if skiprows is not None else 1
 
+            column_names = None
+
+            def get_col_names():
+                nonlocal column_names
+                if column_names is None:
+                    # Using pandas to read the column names
+                    column_names = pandas.read_csv(
+                        kwargs["filepath_or_buffer"], nrows=0, engine="c"
+                    ).columns.tolist()
+                return column_names
+
+            if dtype := kwargs["dtype"]:
+                if isinstance(dtype, dict):
+                    column_types = {c: cls._dtype_to_arrow(t) for c, t in dtype.items()}
+                else:
+                    dtype = cls._dtype_to_arrow(dtype)
+                    column_types = {name: dtype for name in get_col_names()}
+            else:
+                column_types = {}
+
+            if parse_dates := kwargs["parse_dates"]:
+                if isinstance(parse_dates, list) and isinstance(
+                    parse_dates[0], (str, int)
+                ):
+                    # Pandas uses datetime64[ns] dtype for dates.
+                    timestamp_dt = pa.timestamp("ns")
+                    if names and isinstance(parse_dates[0], str):
+                        # If both `names` and column names in `parse_dates` are
+                        # specified, replace the column names with indices.
+                        parse_dates = [names.index(name) for name in parse_dates]
+                    if isinstance(parse_dates[0], int):
+                        # If column indices are specified, load the column names
+                        # with pandas and replace the indices with column names.
+                        column_names = get_col_names()
+                        parse_dates = [column_names[i] for i in parse_dates]
+                    for c in parse_dates:
+                        column_types[c] = timestamp_dt
+                elif not isinstance(parse_dates, bool):
+                    raise NotImplementedError(
+                        f"Argument parse_dates={parse_dates} is not supported"
+                    )
```

Comment:
do we also need to check else condition here?

Reply:
Dict and list of lists are not supported, because we can't combine columns with pyarrow.